### PR TITLE
feat(sdk): expose snapshot diff facade

### PR DESCRIPTION
**Pull request by tjrasche**
## Summary

Expose snapshot comparison through the public `pkg/client/v1` facade using facade-owned result types, while reusing the existing `pkg/diff` implementation. Route `aicr diff` through the SDK facade and preserve its table, JSON, YAML, and drift-gating behavior.

## Motivation / Context

Integrators need programmatic drift detection without importing the compatibility-unstable `pkg/diff` package. This implements the snapshot diff portion of the broader facade-parity issue; mirror and trust remain separate follow-up work.

Fixes: N/A
Related: #2025

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Build/CI/tooling

## Component(s) Affected

- [x] CLI (`cmd/aicr`, `pkg/cli`)
- [ ] API server (`cmd/aicrd`, `pkg/server`)
- [ ] Recipe engine / data (`pkg/recipe`)
- [ ] Bundlers (`pkg/bundler`, `pkg/component/*`)
- [ ] Collectors / snapshotter (`pkg/collector`, `pkg/snapshotter`)
- [ ] Validator (`pkg/validator`)
- [ ] Core libraries (`pkg/errors`, `pkg/k8s`)
- [x] Docs/examples (`docs/`, `examples/`)
- [x] Other: Public Go SDK (`pkg/client/v1`)

## Implementation Notes

- Adds facade-owned change kinds, severity, options, change records, summary, and result types.
- Adds `Client.DiffSnapshots`, `SnapshotDiff.HasDrift`, and `WriteSnapshotDiffTable` without exposing `pkg/diff` types.
- Treats drift as result data rather than an operation error.
- Keeps source labels as output metadata only; labels do not affect comparison semantics.
- Rejects nil, hand-constructed, and wrapped snapshots without a usable typed measurement so malformed inputs cannot produce a false no-drift result.
- Performs the comparison in memory with no new facade timeout or dependency on the OCI recipe-source work.

## Testing

Passed:

```bash
make qualify
go test -race -count=1 ./pkg/client/v1/... ./pkg/cli/...
golangci-lint run -c .golangci.yaml ./pkg/client/v1/... ./pkg/cli/...
API_DIFF_BASELINE=8703d8cfb8dfbbfb3e5016d96c69209a57b19cae ./tools/api-diff
make check-docs-mdx check-docs-mdx-parse
```

Full qualification passed. The API comparison reports compatible additions only.

## Risk Assessment

- [ ] **Low** — Isolated change, well-tested, easy to revert
- [x] **Medium** — Touches multiple components or has broader impact
- [ ] **High** — Breaking change, affects critical paths, or complex rollout

The change is additive and preserves existing CLI output, but it permanently expands the compatibility-reviewed public SDK surface.

**Rollout notes:** N/A; no migration or feature flag is required.

## Checklist

- [x] Tests pass locally (`make test` with `-race`)
- [x] Linter passes (`make lint`)
- [x] I did not skip/disable tests to make CI green
- [x] I added/updated tests for new functionality
- [x] I updated docs if user-facing behavior changed
- [x] Changes follow existing patterns in the codebase
- [x] Commits are cryptographically signed (`git commit -S`) — [GPG signing info](https://docs.github.com/en/authentication/managing-commit-signature-verification)

### CODE REVIEW

File: pkg/client/v1/diff.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	targetInternal, err := requireSnapshotDiffPayload(target, "target")
+	if err != nil {
+		return nil, err
+	}
+	if err := c.assertOpen(); err != nil {
+		return nil, err
+	}
+	if err := snapshotDiffContextError(ctx); err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (tjrasche) [reply pkg/client/v1/diff.go:120]: No change needed here. Each propagated error is already a `pkg/errors` structured error with the correct code and operation-specific context: the payload guards identify baseline or target, `assertOpen` identifies client state, `snapshotDiffContextError` identifies cancellation or deadline expiry, and `diff.WriteTable` already wraps writer failures as snapshot diff table output errors. Wrapping these again would conflict with the repository guidance not to double-wrap errors that already carry the correct code, while wrapping the facade error in `pkg/cli` would also make the thin adapter reinterpret an SDK error. Resolving as not applicable.

File: pkg/client/v1/diff.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+	result := diff.Snapshots(baselineInternal, targetInternal)
+	if err := snapshotDiffContextError(ctx); err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (tjrasche) [reply pkg/client/v1/diff.go]: Deferring this hardening. Normal `LoadSnapshot` inputs are capped at 1 MiB, and `CollectSnapshot` stages through a size-bounded Kubernetes ConfigMap; only an unusually large snapshot constructed directly through `WrapSnapshot` can make the non-interruptible comparison materially delay cancellation. The facade already checks the caller context before and after comparison and while translating results. Making `pkg/diff` context-aware would broaden this deliberately narrow SDK/CLI handoff without a demonstrated production need, so the cost is not justified in this PR. We can revisit if real workloads show comparison latency that makes cancellation responsiveness material.

File: docs/integrator/public-api.md [hunk 3]
```diff
@@ -65,6 +65,8 @@ unrelated exports in their evolving packages remain free to change.
 | Facade symbol | Translates to/from | Notes |
 |---|---|---|
 | `aicr.Snapshot` | `pkg/snapshotter.Snapshot` | **Facade-owned struct**. Public fields are identifying metadata; full measurement payload is preserved in an unexported field for round-trip through `ValidateState`. Obtain one from `Client.LoadSnapshot` (file, URL, or `cm://` ConfigMap) or `Client.CollectSnapshot` (live capture) — neither requires importing `pkg/snapshotter`. `aicr.WrapSnapshot` remains for the narrower case of lifting a `*snapshotter.Snapshot` you already hold from a direct `pkg/snapshotter` call. |
+| `aicr.SnapshotDiff`, `aicr.SnapshotChange`, `aicr.SnapshotDiffSummary`, `aicr.SnapshotDiffOptions` | `pkg/diff` result shapes | **Facade-owned structs** returned by `Client.DiffSnapshots`. They preserve the CLI's JSON/YAML schema without exposing `pkg/diff` types. Drift is data (`SnapshotDiff.HasDrift`), while invalid or payload-less inputs and context cancellation are errors. |
+| `aicr.SnapshotChangeKind`, `aicr.SnapshotChangeSeverity` and their constants | `pkg/diff` enum values | **Facade-owned string enums** for added, removed, modified, and informational changes. |
```
[the remark this answers is not in the document]
Comment (tjrasche) [reply docs/integrator/public-api.md]: Fixed in 09dd9c95. `SnapshotDiffOptions` now has its own row as the input to `Client.DiffSnapshots`; returned result shapes remain grouped separately. `SnapshotChangeKind` and its added/removed/modified values are documented independently from `SnapshotChangeSeverity` and `SnapshotChangeSeverityInfo`.

File: pkg/client/v1/diff.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	}
+	if err := c.assertOpen(); err != nil {
+		return nil, err
+	}
+	if err := snapshotDiffContextError(ctx); err != nil {
+		return nil, err
+	}
+
+	result := diff.Snapshots(baselineInternal, targetInternal)
```
Comment (mchmarny) [inline pkg/client/v1/diff.go]: **P1 — make cancellation effective during the comparison.** This samples `ctx` only before and after `diff.Snapshots`; the comparator sorts and walks all measurements, readings, and items without observing cancellation. `WrapSnapshot` is a documented, unbounded supported input, so a canceled or deadline-expired context can keep the caller blocked until the full CPU traversal completes. The [current cancellation test](https://github.com/NVIDIA/aicr/blob/690dd865c6a2b648ef276a86bfa13446ccb425e5/pkg/client/v1/diff_test.go#L212) only supplies contexts canceled before invocation. Add a context-aware comparison path and a regression that cancels after traversal starts.

File: pkg/client/v1/api-diff-exceptions.yaml [hunk 5]
```diff
@@ -14,4 +14,11 @@
 
 # Intentional API compatibility breaks. Each entry is tied to the active release
 # baseline and must be removed when that baseline advances; stale entries fail the gate.
-acknowledgements: []
+acknowledgements:
+  - baseline: v0.18.0
```
Comment (mchmarny) [inline pkg/client/v1/api-diff-exceptions.yaml]: **P1 — remove the stale v0.18 API-diff acknowledgement.** v0.19.0 is the active stable baseline, and [PR #2167](https://github.com/NVIDIA/aicr/pull/2167) removed exactly this entry from main. With a clean diff the gate only warns, but [tools/api-diff](https://github.com/NVIDIA/aicr/blob/690dd865c6a2b648ef276a86bfa13446ccb425e5/tools/api-diff#L418-L440) hard-exits on any future incompatible SDK change before adjudicating the current acknowledgement. This restores the known stale-baseline trap; keep `acknowledgements: []`.

**Review by mchmarny — CHANGES_REQUESTED** [review]
Requesting changes: two merge blockers remain on exact head `690dd865c6a2b648ef276a86bfa13446ccb425e5`; details are inline. Current CI is complete and green (51 passed, 19 skipped). The branch is currently conflicting with `main`; that mechanical state is separate from this correctness verdict.

**Comment by tjrasche** [comment]
Rebased onto current `main` after #2212 merged and force-pushed to resolve the SDK facade overlap. Original head: `690dd865c6a2b648ef276a86bfa13446ccb425e5`. `main` advanced during post-rebase validation, so the final head is `51ccd6a1ed751fc8715840f52e1bc9280646a2e0`, based on `main` at `21b538df05249f6cf2cbf0aab4dbe9183b045cd3`; the intermediate `681d8664` head is superseded. Reviewers should use the final head; inline anchors from the old history may be outdated.

File: pkg/client/v1/diff.go (hunk 4 above)
Comment (tjrasche) [reply pkg/client/v1/diff.go]: Fixed in `32818788`. `pkg/diff` now exposes a context-aware snapshot comparison path and checks cancellation throughout the existing measurement, topology, subtype, reading, string, item, sort-boundary, and summary traversal. `Client.DiffSnapshots` uses that path and preserves `ErrCodeCanceled` / `ErrCodeTimeout` plus the underlying context cause. Deterministic tests cancel after traversal has accumulated changes and verify that no partial result escapes; the cancellation and legacy-output tests passed 20× under `-race`, and the full focused diff/SDK/CLI race suite passes.

File: pkg/client/v1/api-diff-exceptions.yaml (hunk 5 above)
Comment (tjrasche) [reply pkg/client/v1/api-diff-exceptions.yaml]: Fixed in `32818788`. The stale `v0.18.0` acknowledgement is removed and the file is back to `acknowledgements: []`. The sandbox had been missing the current annotated tag, which caused the earlier incorrect baseline selection. After importing GitHub’s exact signed `v0.19.0` tag object (`7d8fd23e`, peeling to `f1f63463`), `make api-diff` auto-selects `v0.19.0`, exits successfully, and reports no incompatible SDK facade or transparent-alias target changes.

File: pkg/diff/diff.go [hunk 6]
```diff
@@ -110,41 +113,108 @@ func (r *Result) HasDrift() bool {
...
 func Snapshots(baseline, target *snapshotter.Snapshot) *Result {
-	if baseline == nil || target == nil {
+	result, err := snapshots(context.Background(), baseline, target)
+	if err != nil {
+		// context.Background cannot be canceled. Keep the legacy no-error API
+		// total if that invariant ever changes.
 		return &Result{Changes: make([]Change, 0)}
 	}
+	return result
+}
```
[the remark this answers is not in the document]
Comment (tjrasche) [reply pkg/diff/diff.go:123]: Leaving this unchanged. `Snapshots` invokes the private comparator with `context.Background()`, and the comparator currently returns only context errors, so this branch is unreachable under that invariant. Adding production logging to an impossible library fallback would introduce global output/noise. If the comparator later gains another error source, the legacy no-error API behavior should be revisited with that change.

File: pkg/diff/diff.go [hunk 7]
```diff
@@ -330,107 +478,213 @@ func compareItems(prefix string, base, target []measurement.ItemEntry) []Change
...
 	keys := make([]string, 0, len(seen))
 	for k := range seen {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
 		keys = append(keys, k)
 	}
-	return keys
+	return keys, nil
+}
```
[the remark this answers is not in the document]
Comment (tjrasche) [reply pkg/diff/diff.go:683]: Leaving this unchanged. There is no benchmark showing that per-key `ctx.Err()` checks materially affect snapshot diff performance, while striding reduces cancellation granularity and introduces a new cross-package default. That scope is not justified for this narrow facade change; this can be reconsidered with profiling evidence.

File: pkg/diff/diff.go [hunk 8]
```diff
@@ -330,107 +478,213 @@ func compareItems(prefix string, base, target []measurement.ItemEntry) []Change
...
+
+func snapshotContextError(cause error) error {
+	if stderrors.Is(cause, context.Canceled) {
+		return errors.Wrap(errors.ErrCodeCanceled, "snapshot diff canceled", cause)
+	}
+	return errors.Wrap(errors.ErrCodeTimeout, "snapshot diff deadline exceeded", cause)
 }
```
[the remark this answers is not in the document]
Comment (tjrasche) [reply pkg/diff/diff.go:690]: Leaving this unchanged. The `context.Context` contract restricts `Err()` to `nil`, `context.Canceled`, or `context.DeadlineExceeded`; arbitrary values are allowed by `context.Cause`, not `Err`. These helpers classify values obtained from `Err()`, so an internal fallback would only defend contract-violating custom contexts without changing supported behavior.

File: pkg/diff/topology.go [hunk 9]
```diff
@@ -152,24 +204,33 @@ func alignMeasurement(m *measurement.Measurement, plan map[string]subtypePlan) *
...
+				return nil, err
+			}
 			if _, ok := data[key]; ok {
 				data[key] = measurement.Int(count)
 			}
 		}
 		st.Data = data
 	}
-	return &out
+	return &out, nil
```
[the remark this answers is not in the document]
Comment (tjrasche) [reply pkg/diff/topology.go:235]: Leaving this unchanged. The folded map contains at most two entries, and no benchmark indicates that these checks are measurable. Retaining the direct checks keeps cancellation behavior explicit and avoids coupling this loop to the declined striding abstraction.

**Comment by tjrasche** [comment]
History rewrite note: after syncing with current `NVIDIA/main`, I squashed the PR history into one signed, DCO-compliant commit. The old head was `276a019e012f5a2c0ebab098b364741d4e3263d3`; the new head is `91678e9cc0b029b461e1002f75c6dfeada79b5a0`, directly atop `90a7f41e5db5b05c954cffefc62c7e5e500f7eba`. The resulting tree is unchanged (`3669dcac276dfee612625b2c7cfc5bdd395dc74b`). Prior inline anchors may now be outdated; please review from the new head.

File: pkg/client/v1/diff.go [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+// comparison.
+type SnapshotDiffOptions struct {
+	BaselineSource string
+	TargetSource   string
+}
+
+// SnapshotChange is one field-level difference between two snapshots.
+// Baseline and Target are pointers so an absent side remains distinguishable
+// from a present value whose string representation is empty.
+type SnapshotChange struct {
```
Comment (njhensley) [inline pkg/client/v1/diff.go:58]: **🟡 Minor — No JSON/YAML wire-schema golden test for the facade result**

The CLI now serializes *aicr.SnapshotDiff directly, so these facade struct json/yaml tags ARE the SDK's wire contract. stability_test.go pins field names/types via field-access but never marshals and asserts serialized keys or omitempty; cli/diff_test.go's JSON tests assert only drift/exit behavior. A future tag rename or a dropped omitempty on either the facade or pkg/diff would silently change `aicr diff -o json/yaml` output with every test still green.

**Blast radius:** Silent wire-format drift on the compatibility-reviewed SDK surface and the CLI's serialized output — the absent-vs-present-empty (*string + omitempty) distinction is likewise unguarded at the wire level.

**Fix:** Add one golden test that marshals a SnapshotDiff (added / removed / modified-to-empty change) to JSON and YAML and asserts the exact keys — ideally asserting facade output equals pkg/diff-marshaled output so the two tag sets can't diverge. This also guards F3 (a new field added to diff.Change being silently dropped on the facade path).

**Review by njhensley — APPROVED** [review]
## Review — multi-persona + adversarial meta-review

**Method:** 4 parallel persona reviewers (Correctness · SDK/API-contract · Domain & Architecture · CI-DX/Operability) → a senior meta-reviewer re-deriving each finding from the resolved code at `91678e9c`.

**Legend:** 🔴 Blocker · 🟠 Major · 🟡 Minor · 🔵 Nitpick

Clean, well-tested additive SDK change. The facade↔internal translation preserves `aicr diff` output semantics exactly (kinds, severity, paths, summary counts, ordering, and the absent-vs-present-empty `*string` distinction); the new public surface is consistent with existing facade conventions; and the context-cancellation plumbing through `pkg/diff` is real and genuinely tested (mid-traversal cancel asserts a `nil` result and distinct `Canceled`/`Timeout` codes). **No blocker or major findings survived meta-review** — one Minor (inline) and eight Nitpicks, none a live defect introduced by this PR. Approving with comments.

> Note: @mchmarny's earlier **CHANGES_REQUESTED** was on head `690dd865`; the branch has since been rebased to `91678e9c`. Both of those P1 blockers are independently confirmed addressed on the current head (table below).

### Prior-feedback status

| Prior finding | Status | Evidence |
|---|---|---|
| **@mchmarny P1** — cancellation not effective during comparison | ✔️ Addressed | `pkg/diff` now has `SnapshotsWithContext` + `ctx.Err()` threaded through the full traversal; `TestSnapshotsWithContext_MidTraversalCancellation` asserts `result == nil` + correct code + cause |
| **@mchmarny P1** — stale v0.18 api-diff acknowledgement | ✔️ Addressed | `pkg/client/v1/api-diff-exceptions.yaml` is now `acknowledgements: []` (matches main) |
| CodeRabbit — public-api type-role wording | ✔️ Addressed (09dd9c9) | Current table verified correct |
| CodeRabbit — "wrap errors at boundaries" (Minor) | ⊘ Not a violation | Every new `return err` passes through an error that already carries a proper `pkg/errors` code → per the repo's "don't double-wrap errors that already have proper codes" rule, pass-through is correct |

### Findings

**🟡 Minor** — missing JSON/YAML wire-schema golden test → see inline comment on `pkg/client/v1/diff.go`.

**🔵 F1 — one-sided topology hydration failure → asymmetric drop → spurious drift (pre-existing).** `pkg/diff/topology.go:178`: `alignMeasurement` aligns base and target independently with per-side local copies of the plan `p`; if `topology.HydrateItems` fails on exactly one side it sets `p.dropData=false` locally and keeps folded `Data` + un-hydrated `Items`, while the other side has `Data=nil` + hydrated `Items` → `compareMeasurements` emits spurious added/removed changes that could flip `aicr diff --fail-on-drift`. Mechanism confirmed, but this logic is **pre-existing** (this PR only added `ctx` plumbing around it) and reachability is low. Suggest a **separate follow-up issue** rather than gating this PR.

**🔵 F3 — manual facade↔internal struct mirror can silently drop a future field.** `pkg/client/v1/diff.go:143`: result types are hand-mirrored from `pkg/diff` with total string casts. No current loss (only `SeverityInfo` is emitted; severity round-trips). Hazard: a new field added to `diff.Change` would be silently omitted on the facade path with no compiler error. The golden/parity test suggested inline covers this too.

**🔵 F4 — final `ctx.Err()` checks after O(1) statements.** `pkg/diff/diff.go:228-230` (after `Summary.Total`) and `:210-212` (after `sort.Slice`) can only turn an already-complete result into an error on exact-timing cancel — a deliberate fail-closed "no partial result" choice, not a regression. Safe to leave.

**🔵 F5 — redundant belt-and-suspenders `ctx.Err()` checks** bracketing single `sort`/O(1) statements (e.g. `pkg/diff/diff.go:380`); harmless noise. Consider one check per loop iteration + one per O(n log n) sort.

**🔵 F6 — `assertOpen()` ordered after payload validation** (`pkg/client/v1/diff.go:118`): a closed client with a bad snapshot returns the payload error rather than the closed-client error. Both are `ErrCodeInvalidRequest`; cosmetic. Consistent with `LoadSnapshot`'s precedent.

**🔵 F7 — context-error classifier defaults any non-`Canceled` to `Timeout`** (`pkg/diff/diff.go:685`, and the facade's `snapshotDiffContextError`): safe today (`snapshots()` only ever propagates `ctx.Err()`), latent fragility if a non-context error is ever propagated. Consider defaulting `else` to `ErrCodeInternal` + handling `DeadlineExceeded` explicitly.

**🔵 F8 — legacy `Snapshots()` error branch is dead/uncovered** (`pkg/diff/diff.go:116`): `context.Background()` never cancels; documented defensive code, shows as uncovered lines. Acceptable as a totality guard.

**🔵 F9 — `topology.go` `ctx` checkpoints lack a topology-specific cancellation test** (`topology_test.go` untouched, drives uncancelable `context.Background()`). Core loops are covered by the generic mid-traversal tests; the topology-alignment checkpoints are exercised only on their nil-return path. Small additive coverage gap.

### Confirmed non-issues (examined, cleared)
- **Error wrapping** — all new `return err` pass through coded `pkg/errors` errors (correct pass-through, not bare returns).
- **Semantics** — facade→internal copy is 1:1; `WriteTable` reads only fields the facade round-trips, so double-translation drops nothing rendered.
- **Topology alignment logic** — byte-for-byte preserved under `ctx` threading (only `ctx.Err()` insertions + error plumbing added).
- **`copySnapshotDiffString`** correct nil vs `&""` distinction, no aliasing; **`requireSnapshotDiffPayload`** no false accept/reject; **determinism** sorted by `Path` + all map keys, regression-tested; **coverage** no new exported func at 0%; **docs** (go-library.md / public-api.md) accurate incl. the "no facade cap" timeout entry.

```
🔴 Blocker 0 | 🟠 Major 0 | 🟡 Minor 1 | 🔵 Nitpick 8      Recommendation: Approve with comments
```

**Review by mchmarny — APPROVED** [review]
